include/pybind11/pybind11.h

```diff
@@ -1075,15 +1083,27 @@ class class_ : public detail::generic_type {
     }
 
     /// Uses return_value_policy::reference_internal by default
+    template <typename Getter, typename Setter, typename... Extra>
+    class_ &def_property(const char *name, const Getter &fget, const Setter &fset, const Extra& ...extra) {
+        return def_property(name,
+                cpp_function(fget, is_method<type>(*this)), cpp_function(fset, is_method<type>(*this)),
+                return_value_policy::reference_internal, extra...);
+    }
```

Comment:
That's why you also need to change the fully-generic template to only wrap the setter. Let's say that c, d are instances of cpp_function, C=cpp_function, G=Getter (template parameter), S=setter (template parameter) and p, q are non-cpp_function getter/setter. Then the chain of calls for each of the cases is
1. p, q: <G, S> -> <G, C> -> <C, C>
2. c, q: <G, S> -> <C, C>
3. p, c: <G, C> -> <C, C>
4. c, c: <C, C>
I don't know if that notation makes sense. If I've left you just as unclear, let me know and I can send you a PR against your branch.